tests/queries/0_stateless/02523_array_shuffle.sql

```diff
@@ -0,0 +1,67 @@
+SELECT arrayShuffle([]);
+SELECT arrayShuffle([], 0xbad_cafe);
+SELECT arrayShuffle([9223372036854775808]);
+SELECT arrayShuffle([9223372036854775808], 0xbad_cafe);
+SELECT arrayShuffle([1,2,3,4,5,6,7,8,9,10], 0xbad_cafe);
+SELECT arrayShuffle([1,2,3,4,5,6,7,8,9,10.1], 0xbad_cafe);
+SELECT arrayShuffle([1,2,3,4,5,6,7,8,9,9223372036854775808], 0xbad_cafe);
+SELECT arrayShuffle([1,2,3,4,5,6,7,8,9,NULL], 0xbad_cafe);
+SELECT arrayShuffle([toFixedString('123', 3), toFixedString('456', 3), toFixedString('789', 3), toFixedString('ABC', 3), toFixedString('000', 3)], 0xbad_cafe);
+SELECT arrayShuffle([toFixedString('123', 3), toFixedString('456', 3), toFixedString('789', 3), toFixedString('ABC', 3), NULL], 0xbad_cafe);
+SELECT arrayShuffle(['storage','tiger','imposter','terminal','uniform','sensation'], 0xbad_cafe);
+SELECT arrayShuffle(['storage','tiger',NULL,'terminal','uniform','sensation'], 0xbad_cafe);
+SELECT arrayShuffle([NULL]);
+SELECT arrayShuffle([NULL,NULL]);
+SELECT arrayShuffle([[1,2,3,4],[-1,-2,-3,-4],[10,20,30,40],[100,200,300,400,500,600,700,800,900],[2,4,8,16,32,64]], 0xbad_cafe);
+SELECT arrayShuffle([[1,2,3,4],[NULL,-2,-3,-4],[10,20,30,40],[100,200,300,400,500,600,700,800,900],[2,4,8,16,32,64]], 0xbad_cafe);
+SELECT arrayShuffle(groupArray(x),0xbad_cafe) FROM (SELECT number as x from system.numbers LIMIT 100);
+SELECT arrayShuffle(groupArray(toUInt64(x)),0xbad_cafe) FROM (SELECT number as x from system.numbers LIMIT 100);
+SELECT arrayShuffle([tuple(1, -1), tuple(99999999, -99999999), tuple(3, -3)], 0xbad_cafe);
+SELECT arrayShuffle([tuple(1, NULL), tuple(2, 'a'), tuple(3, 'A')], 0xbad_cafe);
```

Comment:
We don't materialize constant in this functions, so result is different for constant and materialized column:

```sql

SELECT
    arrayShuffle(materialize([1, 2, 3]), 42),
    arrayShuffle([1, 2, 3], 42)
FROM numbers(10)

Query id: 6dfe07fe-b4b5-4139-a539-9914c6e1abf5

┌─arrayShuffle(materialize([1, 2, 3]), 42)─┬─arrayShuffle([1, 2, 3], 42)─┐
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,1,2]                                  │ [3,2,1]                     │
│ [1,3,2]                                  │ [3,2,1]                     │
│ [2,1,3]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [1,2,3]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [2,1,3]                                  │ [3,2,1]                     │
└──────────────────────────────────────────┴─────────────────────────────┘
```

This behavior is ok, but should tested as well (and probably documented)

Reply:
Good point. `seed` is being used once for the whole data to be shuffled. We could reseed after each block, **but not sure if this is required (or even intended)**:

```diff
diff --git a/src/Functions/array/arrayShuffle.cpp b/src/Functions/array/arrayShuffle.cpp
index 6532fb89ae..752d4a0ccc 100644
--- a/src/Functions/array/arrayShuffle.cpp
+++ b/src/Functions/array/arrayShuffle.cpp
@@ -102,7 +102,7 @@ public:
     ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t) const override;
 
 private:
-    static ColumnPtr executeGeneric(const ColumnArray & array, pcg64_fast & rng, size_t limit);
+    static ColumnPtr executeGeneric(const ColumnArray & array, uint64_t seed, size_t limit);
 };
 
 template <typename Traits>
@@ -121,7 +121,6 @@ ColumnPtr FunctionArrayShuffleImpl<Traits>::executeImpl(const ColumnsWithTypeAnd
         const auto * val = arguments[Traits::seed_param_idx].column.get();
         return val->getUInt(0);
     }();
-    pcg64_fast rng(seed);
 
     size_t limit = [&]() -> size_t
     {
@@ -136,14 +135,16 @@ ColumnPtr FunctionArrayShuffleImpl<Traits>::executeImpl(const ColumnsWithTypeAnd
         return 0;
     }();
 
-    return executeGeneric(*array, rng, limit);
+    return executeGeneric(*array, seed, limit);
 }
 
 template <typename Traits>
-ColumnPtr FunctionArrayShuffleImpl<Traits>::executeGeneric(const ColumnArray & array, pcg64_fast & rng, size_t limit [[maybe_unused]])
+ColumnPtr FunctionArrayShuffleImpl<Traits>::executeGeneric(const ColumnArray & array, uint64_t seed, size_t limit [[maybe_unused]])
 {
     const ColumnArray::Offsets & offsets = array.getOffsets();
 
+    pcg64_fast rng(seed);
+
     size_t size = offsets.size();
     size_t nested_size = array.getData().size();
     IColumn::Permutation permutation(nested_size);
@@ -166,6 +167,7 @@ ColumnPtr FunctionArrayShuffleImpl<Traits>::executeGeneric(const ColumnArray & a
         else
             shuffle(&permutation[current_offset], &permutation[next_offset], rng);
         current_offset = next_offset;
+        rng.seed(seed);
     }
     return ColumnArray::create(array.getData().permute(permutation, 0), array.getOffsetsPtr());
 }
```

This results in:
```sql
SELECT
    arrayShuffle(materialize([1, 2, 3]), 42),
    arrayShuffle([1, 2, 3], 42)
FROM numbers(10)

Query id: 4a666b06-ecc8-4c30-b8c3-82e3d1fde4ae

┌─arrayShuffle(materialize([1, 2, 3]), 42)─┬─arrayShuffle([1, 2, 3], 42)─┐
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
└──────────────────────────────────────────┴─────────────────────────────
```

Reply:
`arrayShuffle(materialize([1, 2, 3]), 42)` should return different permutations for each row. I meant, for constant array, we can do the same (materialize it automatically), but probably current behavior is even better.